src/field.rs

```diff
@@ -200,78 +200,86 @@ pub trait FieldElementWithInteger: FieldElement + From<Self::Integer> {
     ///
-    /// * `input` - The field element to encode
-    /// * `output` - The slice to write the encoded bits into. Least signicant bit comes first
-    fn fill_with_bitvector_representation(
-        input: &Self::Integer,
-        output: &mut [Self],
-    ) -> Result<(), FieldError> {
-        // Create a mutable copy of `input`. In each iteration of the following loop we take the
-        // least significant bit, and shift input to the right by one bit.
-        let mut i = *input;
-
-        let one = Self::one_integer();
-        for bit in output.iter_mut() {
-            let w = Self::from(i & one);
-            *bit = w;
-            i = i >> 1;
+    /// Returns an error if `input` cannot be represented with `bits` many bits, or if `bits`
+    /// is larger than the bit width of the field's modulus.
+    fn encode_into_bitvector_representation(
+        input: Self::Integer,
+        bits: usize,
+    ) -> Result<BitvectorRepresentationIter<Self>, FieldError> {
+        if !Self::valid_integer_bitlength(bits) {
+            return Err(FieldError::ModulusOverflow);
         }
-
-        // If `i` is still not zero, this means that it cannot be encoded by `bits` bits.
-        if i != Self::zero_integer() {
+        if input >> bits != Self::zero_integer() {
             return Err(FieldError::InputSizeMismatch);
         }
 
-        Ok(())
+        Ok(BitvectorRepresentationIter {
+            inner: 0..bits,
+            input,
+        })
     }
 
-    /// Encode `input` as `bits`-bit vector of elements of `Self` if it's small enough
-    /// to be represented with that many bits.
+    /// Inverts the encoding done by [`Self::encode_into_bitvector_representation`], and returns a
+    /// single field element.
     ///
-    /// # Arguments
+    /// This performs an inner product between the input vector of field elements and successive
+    /// powers of two (starting with 2^0 = 1). If the input came from
+    /// [`Self::encode_into_bitvector_representation`], then the result will be equal to the
+    /// originally encoded integer, projected into the field.
     ///
-    /// * `input` - The field element to encode
-    /// * `bits` - The number of bits to use for the encoding
-    fn encode_into_bitvector_representation(
-        input: &Self::Integer,
-        bits: usize,
-    ) -> Result<Vec<Self>, FieldError> {
-        let mut result = vec![Self::zero(); bits];
-        Self::fill_with_bitvector_representation(input, &mut result)?;
-        Ok(result)
-    }
-
-    /// Decode the bitvector-represented value `input` into a simple representation as a single
-    /// field element.
+    /// Note that this decoding operation is linear, so it can be applied to secret shares of an
+    /// encoded integer, and if the results are summed up, it will be equal to the encoded integer.
     ///
     /// # Errors
     ///
     /// This function errors if `2^input.len() - 1` does not fit into the field `Self`.
```

Comment:
nit: IIUC this condition is symmetric to "Returns an error if `input` cannot be represented with `bits` many bits, or if `bits` is larger than the bit width of the field's modulus" on `encode_into_bitvector_representation`, and it'd be nice if they were written in the same terms (e.g. relative to the width of the modulus of the field).